src/ui/public/agg_types/buckets/terms_other_bucket_helper.js

```diff
@@ -0,0 +1,143 @@
+        return bucketKey === keyParts[i];
+      });
+      if (bucket) {
+        responseAgg = bucket;
+      }
+    }
+  }
+  if (responseAgg[aggWithOtherBucket.id]) return responseAgg[aggWithOtherBucket.id].buckets;
+  return [];
+};
+
+const getAggConfigResult = (responseAggs, aggId, bucketKey) => {
+  let resultBuckets = [];
+  if (responseAggs[aggId]) {
+    const matchingBucket = responseAggs[aggId].buckets.find(bucket => bucket.key === bucketKey);
+    if (matchingBucket) resultBuckets.push(matchingBucket);
+    return resultBuckets;
+  }
+  _.each(responseAggs, agg => {
+    resultBuckets = [
+      ...resultBuckets,
+      ...getAggConfigResult(agg.aggs, aggId, bucketKey)
+    ];
+  });
+
+  return resultBuckets;
+};
+
+
+export const OtherBucketHelperProvider = (Private) => {
+  const AggConfig = Private(VisAggConfigProvider);
+
+  const buildOtherBucketAgg = (aggsConfig, aggs, aggWithOtherBucket, response) => {
+    const bucketAggs = aggsConfig.filter(agg => agg.type.type === 'buckets');
+    const index = bucketAggs.findIndex(agg => agg.id === aggWithOtherBucket.id);
+
+    // create filters aggregation
+    const filterAgg = new AggConfig(aggsConfig[index].vis, {
+      type: 'filters',
+      id: 'other',
+      schema: {
+        group: 'buckets'
+      }
+    });
+
+    // nest all the child aggregations of aggWithOtherBucket
+    const filterAggDsl = filterAgg.toDsl();
+    const resultAgg = {
+      aggs: getAggConfig(aggs, aggWithOtherBucket).aggs,
+      filters: filterAggDsl
+    };
+
+    // create filters for all parent aggregation buckets
+    const walkBucketTree = (aggIndex, aggs, aggId, filters, key) => {
+      const agg = aggs[aggId];
+      const newAggIndex = aggIndex + 1;
+      const newAgg = bucketAggs[newAggIndex];
+      const currentAgg = bucketAggs[aggIndex];
+      if (aggIndex < index) {
+        _.each(agg.buckets, (bucket, bucketObjKey) => {
+          const bucketKey = currentAgg.getKey(bucket, Number.isInteger(bucketObjKey) ? null : bucketObjKey);
+          const filter = _.cloneDeep(bucket.filter) || currentAgg.createFilter(bucketKey);
+          delete filter.meta;
+          const migratedFilter = migrateFilter(filter.query || filter);
+          const newFilters = [...filters, migratedFilter];
+          walkBucketTree(newAggIndex, bucket, newAgg.id, newFilters, `${key}-${bucketKey.toString()}`);
+        });
+        return;
+      }
+
+      filterAggDsl.filters[key] = {
+        bool: { must: filters, must_not: [ { terms: {} } ] }
+      };
+
+      if (agg.buckets.find(bucket => bucket.key === '__missing__')) {
+        filterAggDsl.filters[key].bool.must.push({
+          exists: {
+            field: aggWithOtherBucket.params.field.name
+          }
+        });
+      }
+
+      // create not filters for all the buckets
+      const notKeys = agg.buckets.map(bucket => bucket.key);
+      filterAggDsl.filters[key].bool.must_not[0].terms[aggWithOtherBucket.params.field.name] = notKeys;
```

Comment:
I don't think you want to use a `terms` query. If a user is aggregating on an analyzed field (with field data turned on) these term queries won't match properly. Scripted fields also won't work with the current implementation. I think a negated [phrases](https://github.com/elastic/kibana/blob/013b7943dc8358f29c6e52dbefad02b9cefe8fc6/src/ui/public/filter_manager/lib/phrases.js#L3) filter would work. It'll create match_phrase queries, so my only question would be whether the terms agg supports any data types that won't work with match_phrase. 

It could also make more sense to use the agg's createFilter method, that way you know you're getting a valid filter for this particular agg and field type. It already uses a filter constructor under the hood. Negate them all and then combine them with a bool using the `buildQueryFromFilters` helper I mentioned above.

(as a general rule, if you ever find yourself manually building query DSL in your code, there's a good chance you should be using one of our filter/query abstractions instead)